### PR TITLE
Upgrade to arkworks 0.4.2 (mina compatible / p-s develop)

### PR DESCRIPTION
**Pull request by volhovm**
Rebasing the epic PR now on top of the most recent proof-system's `develop` which maps to mina's `compatible`.

### CODE REVIEW

**Comment by dannywillems** [comment]
Oh boy :fearful:

File: Cargo.lock [hunk 1]
```diff
@@ -48,21 +59,29 @@ dependencies = [
...
+ "num-integer",
+ "num-traits",
+ "serde",
+ "serde_derive",
+ "serde_json",
+ "sha2",
 ]
 
 [[package]]
 name = "ark-bn254"
```
Comment (dannywillems) [inline Cargo.lock:81]: Next step is to get rid of this. It is not in arkworks. Opening follow-up issue.

File: circuit-construction/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,47 @@
+[package]
```
Comment (dannywillems) [inline circuit-construction/Cargo.toml:1]: Where does it come from?

File: tools/kimchi-visu/Cargo.toml [hunk 3]
```diff
@@ -13,8 +13,8 @@ license = "Apache-2.0"
 path = "src/lib.rs"
 
 [dependencies]
-ark-ec = "0.3.0"
-ark-ff = "0.3.0"
+ark-ec = "0.4.2"
```
Comment (dannywillems) [inline tools/kimchi-visu/Cargo.toml:16]: Note: would be good to backport also the workspace usage as we do have in master.

File: curves/src/pasta/fields/mod.rs [hunk 4]
```diff
@@ -1,8 +1,48 @@
...
 pub mod fp;
 pub use self::fp::*;
 
 pub mod fq;
 pub use self::fq::*;
 
+pub mod fft;
+
+#[derive(Debug, PartialEq)]
+pub enum LegendreSymbol {
```
Comment (dannywillems) [inline curves/src/pasta/fields/mod.rs:11]: cool

File: curves/src/pasta/curves/tests.rs [hunk 5]
```diff
@@ -1,28 +1,6 @@
...
-    group_test(a, b);
-}
-
-#[test]
-fn test_pallas_generator() {
-    let generator = pallas::Pallas::prime_subgroup_generator();
-    assert!(generator.is_on_curve());
-    assert!(generator.is_in_correct_subgroup_assuming_on_curve());
-}
+test_group!(g1; ProjectivePallas; sw);
```
Comment (dannywillems) [inline curves/src/pasta/curves/tests.rs:5]: Cool, more tests

File: Cargo.lock [hunk 6]
```diff
@@ -48,21 +59,29 @@ dependencies = [
...
-version = "0.3.0"
+version = "0.4.0"
```
Comment (dannywillems) [inline Cargo.lock:82]: I open https://github.com/arkworks-rs/algebra/pull/856 upstream to homogenize the version. For 0.4.3 I hope it will be ok.

**Review by dannywillems — COMMENTED** [review]
LGTM, modulo comment. Not approving because of circuit-construction.

File: circuit-construction/Cargo.toml (hunk 2 above)
Comment (volhovm) [reply circuit-construction/Cargo.toml:1]: I believe this is something @mrmr1993 added when he was working on an earlier version of this PR. This existed in the previous PR when I picked up the task :shrug:

File: tools/kimchi-visu/Cargo.toml (hunk 3 above)
Comment (volhovm) [reply tools/kimchi-visu/Cargo.toml:16]: Yes! The way it is now in compatible is very annoying. But I don't want to have more conflicts as of now, so can do it later.

File: Cargo.lock (hunk 6 above)
Comment (dannywillems) [reply Cargo.lock:82]: Opening follow-up https://github.com/o1-labs/proof-systems/issues/2555

**Review by dannywillems — APPROVED** [review]
LGTM

**Pull request merged**